server/src/main/java/org/apache/iotdb/db/conf/adapter/ActiveTimeSeriesCounter.java

```diff
@@ -0,0 +1,145 @@
+  private static Map<String, HyperLogLog> storageGroupHllMap = new ConcurrentHashMap<>();
+
+  /**
+   * Map[StorageGroup, ActiveTimeSeriesRatio]
+   */
+  private static Map<String, Double> activeRatioMap = new ConcurrentHashMap<>();
+
+  /**
+   * Map[StorageGroup, ActiveTimeSeriesNumber]
+   */
+  private static Map<String, Long> activeTimeSeriesNumMap = new ConcurrentHashMap<>();
+
+  /**
+   * LOG2M decide the precision of the HyperLogLog algorithm
+   */
+  static final int LOG2M = 13;
+
+  private ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+
+  @Override
+  public void init(String storageGroup) {
+    storageGroupHllMap.put(storageGroup, new HyperLogLog(LOG2M));
+    activeRatioMap.put(storageGroup, 0D);
+    activeTimeSeriesNumMap.put(storageGroup, 0L);
+  }
+
+  @Override
+  public void offer(String storageGroup, String device, String measurement) {
+    String path = device + "." + measurement;
+    try {
+      storageGroupHllMap.get(storageGroup).offer(path);
+    } catch (Exception e) {
+      storageGroupHllMap.putIfAbsent(storageGroup, new HyperLogLog(LOG2M));
+      storageGroupHllMap.get(storageGroup).offer(path);
+      LOGGER.error("Register active time series root.{}.{}.{} failed", storageGroup, device,
+          measurement, e);
+    }
+  }
+
+  @Override
+  public void updateActiveRatio(String storageGroup) {
+    lock.writeLock().lock();
+    try {
+      long activeTimeSeriesNum = storageGroupHllMap.get(storageGroup).cardinality();
+      if (activeTimeSeriesNum != activeTimeSeriesNumMap.get(storageGroup)) {
+        // update the active time series number in the newest memtable to be flushed
+        activeTimeSeriesNumMap.put(storageGroup, activeTimeSeriesNum);
+
+        double totalActiveTsNum = 0;
+        LOGGER.debug("{}: updating active ratio", Thread.currentThread().getName());
+        for (double number : activeTimeSeriesNumMap.values()) {
+          totalActiveTsNum += number;
+        }
+        for (Map.Entry<String, Long> entry : activeTimeSeriesNumMap.entrySet()) {
+          double activeRatio = 0;
+          if (totalActiveTsNum > 0) {
+            activeRatio = entry.getValue() / totalActiveTsNum;
+          }
+          activeRatioMap.put(entry.getKey(), activeRatio);
+          LOGGER.debug("{}: storage group {} has an active ratio: {}",
+              Thread.currentThread().getName(),
+              entry.getKey(), activeRatio);
+        }
+      }
+      // initialize the HLL counter
+      storageGroupHllMap.put(storageGroup, new HyperLogLog(LOG2M));
+    } catch (Exception e) {
+      LOGGER.error("Update {} active ratio failed", storageGroup, e);
+    } finally {
+      lock.writeLock().unlock();
+    }
+  }
+
+  @Override
+  public double getActiveRatio(String storageGroup) {
+    lock.writeLock().lock();
+    double ratio;
+    try {
+      ratio = activeRatioMap.get(storageGroup);
+    } catch (Exception e) {
+      LOGGER.error("Get {} active ratio failed", storageGroup, e);
```

Comment:
I don't think it's elegant if we just do `try-catch` in lots of places just in case `unexpected test class`.

Reply:
Fixed